pkg/api/middleware/authorization.go

```diff
@@ -0,0 +1,191 @@
+func NewTagAuthorizer(db storage.Storage, log echo.Logger) *TagAuthorizer {
+	return &TagAuthorizer{
+		db,
+		log,
+	}
+}
+
+// Authorize authorizes HTTP request by verifying resource ownership.
+// It does so by retrieving the tag included in the http request's
+// authorization header and comparing it with the tags associated
+// with the resource that the request is trying to modify.
+// If request tag is among the ones associated with the resource,
+// then action is granted. Otherwise it is denied.
+func (a *TagAuthorizer) Authorize(r *http.Request) error {
+	if !isModificationMethod(r.Method) {
+		// Only modification methods
+		// require authorization
+		return nil
+	}
+
+	authTag := parseAuthTag(r.Header)
+	if authTag == "" {
+		return errUnauthorized
+	}
+
+	// Only makes sense to authorize modifications of
+	// Findings and Assets as these are the entities
+	// that are related with a team
+	if isFindingOp(r.URL.Path) {
+		findingID := parseEntityID(r.URL.Path)
+		return a.authorizeFindingOp(findingID, authTag)
+	}
+	if isTargetOp(r.URL.Path) {
+		targetID := parseEntityID(r.URL.Path)
+		return a.authorizeTargetOp(targetID, authTag)
+	}
+
+	// Deny by default
+	return errUnauthorized
+}
+
+func (a *TagAuthorizer) authorizeFindingOp(findingID, tag string) error {
+	f, err := a.db.GetFinding(findingID)
+	if err != nil {
+		a.log.Errorf("unexpected error authorizing finding op: %v", err)
+		return err
+	}
+
+	if f.ID == "" {
+		return errNotFound
+	}
+
+	if !contains(f.Target.Tags, tag) {
+		return errForbidden
+	}
+
+	return nil
+}
+
+func (a *TagAuthorizer) authorizeTargetOp(targetID, tag string) error {
+	t, err := a.db.GetTarget(targetID)
+	if err != nil {
+		a.log.Errorf("unexpected error authorizing target op: %v", err)
+		return err
+	}
+
+	if t.ID == "" {
+		return errNotFound
+	}
+
+	if !contains(t.Tags, tag) {
+		return errForbidden
+	}
+
+	return nil
+}
+
+func isFindingOp(path string) bool {
+	return strings.HasPrefix(path, findingsOpPrefix)
+}
+
+func isTargetOp(path string) bool {
+	return strings.HasPrefix(path, targetsOpPrefix)
+}
+
+func parseAuthTag(headers http.Header) string {
+	authH := headers.Get("authorization")
+	if authH == "" {
+		return ""
+	}
+	if !strings.HasPrefix(authH, tagAuthSchemaPrefix) {
+		return ""
+	}
+	return strings.TrimPrefix(authH, tagAuthSchemaPrefix)
+}
+
+func parseEntityID(path string) string {
+	if pathParts := strings.Split(path, "/"); len(pathParts) > 1 {
+		return pathParts[2]
+	}
+	return ""
+}
+
+func errToStatus(err error) int {
+	var status int
+
+	if err == nil {
+		status = http.StatusOK
+	} else if errors.Is(err, errUnauthorized) {
+		status = http.StatusUnauthorized
+	} else if errors.Is(err, errForbidden) {
+		status = http.StatusForbidden
+	} else if errors.Is(err, errNotFound) {
+		status = http.StatusNotFound
+	} else {
+		status = http.StatusInternalServerError
+	}
+
+	return status
+}
+
+func isModificationMethod(method string) bool {
+	return method == http.MethodPut ||
```

Comment:
What happens if the user tries to perform a POST? Or we won't allow that for now?